src/bin/wasmtime.rs

```diff
@@ -1,457 +1,81 @@
-    flag_debug: bool,
-    flag_g: bool,
-    flag_enable_simd: bool,
-    flag_lightbeam: bool,
-    flag_cranelift: bool,
-    flag_invoke: Option<String>,
-    flag_preload: Vec<String>,
-    flag_env: Vec<String>,
-    flag_dir: Vec<String>,
-    flag_mapdir: Vec<String>,
-    flag_wasi_c: bool,
-}
-
-fn compute_preopen_dirs(flag_dir: &[String], flag_mapdir: &[String]) -> Vec<(String, File)> {
-    let mut preopen_dirs = Vec::new();
-
-    for dir in flag_dir {
-        let preopen_dir = preopen_dir(dir).unwrap_or_else(|err| {
-            println!("error while pre-opening directory {}: {}", dir, err);
-            exit(1);
-        });
-        preopen_dirs.push((dir.clone(), preopen_dir));
-    }
-
-    for mapdir in flag_mapdir {
-        let parts: Vec<&str> = mapdir.split("::").collect();
-        if parts.len() != 2 {
-            println!(
-                "--mapdir argument must contain exactly one double colon ('::'), separating a \
-                 guest directory name and a host directory name"
-            );
-            exit(1);
-        }
-        let (key, value) = (parts[0], parts[1]);
-        let preopen_dir = preopen_dir(value).unwrap_or_else(|err| {
-            println!("error while pre-opening directory {}: {}", value, err);
-            exit(1);
-        });
-        preopen_dirs.push((key.to_string(), preopen_dir));
-    }
-
-    preopen_dirs
-}
-
-/// Compute the argv array values.
-fn compute_argv(argv0: &str, arg_arg: &[String]) -> Vec<String> {
-    let mut result = Vec::new();
-
-    // Add argv[0], which is the program name. Only include the base name of the
-    // main wasm module, to avoid leaking path information.
-    result.push(
-        Path::new(argv0)
-            .components()
-            .next_back()
-            .map(Component::as_os_str)
-            .and_then(OsStr::to_str)
-            .unwrap_or("")
-            .to_owned(),
-    );
-
-    // Add the remaining arguments.
-    for arg in arg_arg {
-        result.push(arg.to_owned());
-    }
-
-    result
-}
-
-/// Compute the environ array values.
-fn compute_environ(flag_env: &[String]) -> Vec<(String, String)> {
-    let mut result = Vec::new();
-
-    // Add the environment variables, which must be of the form "key=value".
-    for env in flag_env {
-        let split = env.splitn(2, '=').collect::<Vec<_>>();
-        if split.len() != 2 {
-            println!(
-                "environment variables must be of the form \"key=value\"; got \"{}\"",
-                env
-            );
+//! Primarily used to run WebAssembly modules.
+//! See `wasmtime --help` for usage.
+
+use anyhow::Result;
+use clap::{App, AppSettings, ArgMatches, ErrorKind};
+use wasmtime_cli::commands;
+
+type CommandFactory = fn() -> App<'static, 'static>;
+type CommandExecutor = fn(&ArgMatches<'static>) -> Result<()>;
+
+// Edit this list to add new subcommands to the `wasmtime` tool.
+const SUBCOMMANDS: &[(&'static str, CommandFactory, CommandExecutor)] = &[
```

Comment:
This is a nice way to separate concerns :+1: 

I think it may make sense to take this another step further and go beyond `clap`'s stringly-typed `ArgMatches` and use `structopt` (a custom derive for `clap`) so that we parse CLI args directly into structs and enums. 

I've had really nice experiences with this, since now you can implement methods on those structs/enums and you get this separation that you're enforcing via `SUBCOMMANDS` here by design. For example:

* https://github.com/fitzgen/fart/blob/a9a2a7dcb2cb2c646a69a6373ee11fd6afa26075/crates/cli/src/main.rs#L20-L40
* https://github.com/fitzgen/fart/blob/a9a2a7dcb2cb2c646a69a6373ee11fd6afa26075/crates/cli/src/main.rs#L20-L40

> Aside, [this comment](https://github.com/fitzgen/fart/blob/a9a2a7dcb2cb2c646a69a6373ee11fd6afa26075/crates/cli/src/main.rs#L63) is not true, I was ignorant of [the `last` method](https://docs.rs/clap/2.33.0/clap/struct.Arg.html#method.last) at the time, so don't let that hack color your opinion.

I just did this migration with `cargo fuzz`, if you'd like to see that to get an idea of what moving from `clap` to `structopt` is like: https://github.com/rust-fuzz/cargo-fuzz/commit/9804feee14c81eeb895b48f910a3a6e9059e84f4

Reply:
Nice, that looks pretty clean.  Let me take a stab at adapting this to use `structopt`.